### PR TITLE
Remove -XX:+EnableValhalla and -XX:-EnableValhalla

### LINKED ISSUE #14566: Segfault from instanceOfOrCheckCast in many tests running with -XX:-EnableValhalla built from openj9-openjdk-jdk.valuetypes
In running sanity.functional tests on a build using openj9-openjdk-jdk.valuetypes, configured with `--enable-inline-types`, but run with `EXTRA_OPTIONS="-XX:-EnableValhalla"`, most tests failed with a segmentation fault with the following stack backtrace

```
----------- Stack Backtrace -----------
instanceOfOrCheckCast+0x9 (0x00007F88AD6488D9 [libj9vm29.so+0x1528d9])
resolveClassRef+0x619 (0x00007F88AD56D5E9 [libj9vm29.so+0x775e9])
_ZN32VM_BytecodeInterpreterCompressed3runEP10J9VMThread+0x11a05 (0x00007F88AD59FC75 [libj9vm29.so+0xa9c75])
bytecodeLoopCompressed+0x95 (0x00007F88AD58D9B5 [libj9vm29.so+0x979b5])
 (0x00007F88AD6457E2 [libj9vm29.so+0x14f7e2])
---------------------------------------
```

Looking at a core file in gdb, the `castClass` appears to be NULL, in case that's relevant:

```
#13 0x00007f88ad6488d9 in VM_VMHelpers::isSameOrSuperclass (subClass=0x33500, superClass=0x0)
    at /root/hostdir/value-types/openj9-openjdk-jdk/openj9/runtime/oti/VMHelpers.hpp:539
#14 VM_VMHelpers::inlineCheckCast (updateCache=true, castClass=0x0, instanceClass=0x33500)
    at /root/hostdir/value-types/openj9-openjdk-jdk/openj9/runtime/oti/VMHelpers.hpp:569
#15 instanceOfOrCheckCast (instanceClass=0x33500, castClass=0x0) at /root/hostdir/value-types/openj9-openjdk-jdk/openj9/runtime/util/checkcast.cpp:30
#16 0x00007f88ad56d5e9 in resolveClassRef (vmStruct=0x16f00, ramCP=0x343c0, cpIndex=<optimized out>, resolveFlags=65)
    at /root/hostdir/value-types/openj9-openjdk-jdk/openj9/runtime/vm/resolvesupport.cpp:357
#17 0x00007f88ad59fc75 in VM_BytecodeInterpreterCompressed::newLogic (newObject=<optimized out>, _pc=<optimized out>, _sp=<optimized out>, 
    this=<optimized out>) at /root/hostdir/value-types/openj9-openjdk-jdk/openj9/runtime/vm/BytecodeInterpreter.hpp:7808
#18 VM_BytecodeInterpreterCompressed::newdup (_pc=<optimized out>, _sp=<optimized out>, this=<optimized out>)
    at /root/hostdir/value-types/openj9-openjdk-jdk/openj9/runtime/vm/BytecodeInterpreter.hpp:7841
#19 VM_BytecodeInterpreterCompressed::run (this=0x7f88af70fc40, vmThread=0xfff00618)
    at /root/hostdir/value-types/openj9-openjdk-jdk/openj9/runtime/vm/BytecodeInterpreter.hpp:11135
```

A [link to a failing test run with `TARGET=StringIdentTest_0 EXTRA_OPTIONS=-XX:-EnableValhalla`](https://interrnal-server/job/Test_openjdknext_j9_sanity.functional_x86-64_linux_vt_standard_Personal_testList_0/137/) can be found on an internal server.

### PR DESCRIPTION
**Pull request by hangshao0**
1. Remove options -XX:+EnableValhalla and -XX:-EnableValhalla
2. Remove flags related to these 2 options
3. areValueTypesEnabled() return true in a VT build, false otherwise.

closes #14566

Signed-off-by: Hang Shao <hangshao@ca.ibm.com>

### CODE REVIEW

**Comment by hangshao0** [comment]
The segfault goes away with this change.

**Comment by gacholio** [comment]
This change won't work properly because there's an exception left pending during the call to resolve the known class.

**Comment by gacholio** [comment]
Do we know if LinkageError is loaded naturally during startup? If so, we should just preload the known class.

**Comment by gacholio** [comment]
This is another case that would benefit from something we discussed long ago - tracking known classes when they are loaded by java code.

**Comment by gacholio** [comment]
For the record, I expect this change works because LinkageError has already been loaded, so the known class resolve does not need to run any java code (it just finds the entry in the boot loader hash table).

**Comment by gacholio** [comment]
ExceptionInInitializerError is a subclass of LinkageError, and it has a known field, so it's preloaded at startup.

So, there's no reason at all not to also preload LinkageError since it's certainly going to be loaded indirectly anyway.

**Comment by gacholio** [comment]
LinkageError should be added here and the use of the OR_NULL macro can remain unchanged:
https://github.com/eclipse-openj9/openj9/blob/84e6f5ffa8879c1ea9be5c16da64a9baa2abe5e4/runtime/jcl/common/jclcinit.c#L579-L596

**Comment by hangshao0** [comment]
I see `LinkageError` is already there in the `requiredClasses[]` at line 586 of jclcinit.c.

It failed very early in the startup, before we load the requiredClasses in `initializeRequiredClasses()`. (I don't see `requiredClasses[0]` java.lang.Thread been loaded)

```
class load: java/lang/Object from: jrt:/java.base
class load: java/lang/IdentityObject from: jrt:/java.base
class load: java/lang/J9VMInternals from: jrt:/java.base
class load: com/ibm/oti/vm/VM from: jrt:/java.base
class load: java/lang/J9VMInternals$ClassInitializationLock from: jrt:/java.base
class load: java/io/Serializable from: jrt:/java.base
class load: java/lang/reflect/AnnotatedElement from: jrt:/java.base
class load: java/lang/reflect/GenericDeclaration from: jrt:/java.base
class load: java/lang/reflect/Type from: jrt:/java.base
class load: java/lang/constant/Constable from: jrt:/java.base
class load: java/lang/invoke/TypeDescriptor from: jrt:/java.base
class load: java/lang/invoke/TypeDescriptor$OfField from: jrt:/java.base
class load: java/lang/Class from: jrt:/java.base
class load: java/lang/Cloneable from: jrt:/java.base
class load: java/lang/invoke/MethodHandle from: jrt:/java.base
class load: java/lang/Comparable from: jrt:/java.base
class load: java/lang/CharSequence from: jrt:/java.base
class load: java/lang/constant/ConstantDesc from: jrt:/java.base
class load: java/lang/String from: jrt:/java.base
class load: com/ibm/jit/JITHelpers from: jrt:/java.base
class load: jdk/internal/misc/Unsafe from: jrt:/java.base
class load: java/util/Objects from: jrt:/java.base
class load: jdk/internal/reflect/Reflection from: jrt:/java.base
class load: java/lang/Iterable from: jrt:/java.base
class load: java/util/Collection from: jrt:/java.base
class load: java/util/Set from: jrt:/java.base
class load: java/lang/Throwable from: jrt:/java.base
class load: java/lang/Error from: jrt:/java.base
class load: java/lang/LinkageError from: jrt:/java.base
class load: java/lang/ClassFormatError from: jrt:/java.base
class load: java/lang/StackTraceElement from: jrt:/java.base
class load: java/util/Collections from: jrt:/java.base
Unhandled exception
Type=Segmentation error vmState=0x00000000
J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
```

**Comment by hangshao0** [comment]
Updated to use `internalFindKnownClass()` in the error case.

**Comment by gacholio** [comment]
Is this worth fixing at all? I gather from the issue that this is running a VT build with an essentially unsupported command line option. With this fix, what's the behaviour of the VM? I assume it still won't actually get anywhere.

**Comment by hangshao0** [comment]
>With this fix, what's the behaviour of the VM? I assume it still won't actually get anywhere.

The JVM will failed to startup throwing a class format error, instead of segfault. 

> I gather from the issue that this is running a VT build with an essentially unsupported command line option.

The context is our VT JDK is now built on our internal fork of the OpenJDK Valhalla repo, where there are bootstrap classes with VT features. Disabling VT with -XX:-EnableValhalla does not make much sense anymore. 

> Is this worth fixing at all? 

I am fine with not fixing this. If so, we'd better remove `-XX:-EnableValhalla`. 

@tajila are you fine with removing `-XX:-EnableValhalla` ?

**Comment by tajila** [comment]
> @tajila are you fine with removing -XX:-EnableValhalla ?

yes we can remove it, and make the `areValueTypesEnabled` call return true if the jdk support VTs

**Comment by hangshao0** [comment]
> I am fine with not fixing this. If so, we'd better remove -XX:-EnableValhalla.
> @tajila are you fine with removing -XX:-EnableValhalla ?

A bit update from the JIT side https://github.com/eclipse-openj9/openj9/issues/14566#issuecomment-1057337424. They need `-XX:-EnableValhalla` for builds on non-valhalla repo https://github.com/ibmruntimes/openj9-openjdk-jdk.valuetypes.git

**Comment by tajila** [comment]
Update: After todays call we agreed that the `-XX:[-|+]EnableValhalla` options can be removed. The VT enable builds will always return TRUE in `areValueTypesEnabled` and all other builds will return FALSE. The JIT team can use the jdknext builds to compare against VT builds.

**Comment by hangshao0** [comment]
Convert this to draft as we are changing the solution.

**Comment by hangshao0** [comment]
This change has been updated to remove `-XX:+/-EnableValhalla` and it is ready to be reviewed. 

There is one JIT change in `J9ObjectModel.cpp`, @0xdaryl or @hzongaro Can you review that ?

**Review by 0xdaryl — APPROVED** [review]
Approve JIT changes.

File: runtime/oti/j9nonbuilder.h [hunk 1]
```diff
@@ -2025,7 +2025,7 @@ typedef struct J9BCTranslationData {
 #define BCT_Xfuture  0x800
 #define BCT_AlwaysSplitBytecodes 0x1000
 #define BCT_IntermediateDataIsClassfile  0x2000
-#define BCT_ValueTypesEnabled  0x4000
+/* Bit 0x4000 is free to use. */
```
Comment (tajila) [inline runtime/oti/j9nonbuilder.h:2028]: `BCT_Unused_0x4000`

**Review by tajila — APPROVED** [review]

File: runtime/oti/j9consts.h [hunk 2]
```diff
@@ -332,7 +332,7 @@ extern "C" {
 #define J9_EXTENDED_RUNTIME_REDUCE_CPU_MONITOR_OVERHEAD 0x80000000
 
 /* constants for J9JavaVM.extendedRuntimeFlags2 */
-#define J9_EXTENDED_RUNTIME2_ENABLE_VALHALLA 0x1
+/* Bit 0x1 is free to use. */
```
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/oti/j9consts.h:335]: Fixed.

File: runtime/oti/j9nonbuilder.h (hunk 1 above)
Comment (hangshao0) [reply runtime/oti/j9nonbuilder.h:2028]: Fixed.

**Review by gacholio — APPROVED** [review]

**Comment by tajila** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Comment by tajila** [comment]
Jenkins test sanity win jdk8

**Comment by tajila** [comment]
Jenkins test sanity xlinuxvalst jdknext

**Pull request merged**